csrc/scheduler/mma_utils.cpp

```diff
@@ -1437,103 +1416,209 @@ TensorView* getTensorviewPriorToCast(TensorView* in) {
-        {dynamic_cast<BroadcastOp*>(a->definition()),
-         dynamic_cast<BroadcastOp*>(b->definition())}};
-    return props;
+char dtypeToChar(const DataType& dtype) {
+  if (dtype == DataType::Half) {
+    return 'H';
+  } else if (dtype == DataType::BFloat16) {
+    return 'T';
+  } else if (dtype == DataType::Float) {
+    return 'S';
+  } else if (dtype == DataType::Double) {
+    return 'D';
   }
-  return std::nullopt;
+  NVF_ERROR(false, "Unsupported dtype for matmul: ", dtype);
+  return 0;
 }
-} // namespace
 
-void CombineMulSum::handle(ReductionOp* stmt) {
-  // Check if operation is a sum.
-  if (stmt->getReductionOpType() == BinaryOpType::Add) {
-    auto* inputOfSum = stmt->in();
-    if (inputOfSum != nullptr) {
-      auto* expr = inputOfSum->definition();
-      // Then check if the prodcer of the sum is a mul.
-      if (auto bOp = dynamic_cast<BinaryOp*>(expr)) {
-        // If it'a mul followed by a sum, put this in a list.
-        if (bOp->getBinaryOpType() == BinaryOpType::Mul) {
-          // If the Mul-Sum is a valid representation of a matmul,
-          // then get the properties of the replacement Mma op.
-          auto props = getMulSumInsOutsBcasts(bOp, stmt);
-          if (props.has_value()) {
-            mul_sum_props_.push_back(*props);
+namespace {
+
+class MatmulPatternMatcher : IterVisitor {
+ public:
+  static std::vector<MatmulPattern> run(Fusion* fusion) {
+    MatmulPatternMatcher matcher;
+    matcher.traverse(fusion);
+    return matcher.patterns_;
+  }
+
+ private:
+  using IterVisitor::handle;
+
+  // Handle the case when no translation is needed.
+  void handle(MmaOp* mop) override {
+    MatmulPattern& pattern = patterns_.emplace_back();
+    pattern.A = mop->inA()->as<TensorView>();
+    pattern.B = mop->inB()->as<TensorView>();
+    pattern.output = mop->out()->as<TensorView>();
+  }
+
+  void handle(ReductionOp* rop) override {
+    // Check if operation is a sum.
+    if (rop->getReductionOpType() != BinaryOpType::Add) {
+      return;
+    }
+    // Then check if the producer of the sum is a mul.
+    if (auto bop = dynamic_cast<BinaryOp*>(rop->in()->definition())) {
+      if (bop->getBinaryOpType() != BinaryOpType::Mul) {
+        return;
+      }
+      // TODO: Allow multiple K dimensions
+      // Check that there's a single K dimension
+      bool has_k = false;
+      for (IterDomain* id :
+           rop->out()->as<TensorView>()->getMaybeRFactorDomain()) {
+        if (id->isReduction()) {
+          if (has_k) {
+            return;
+          }
+          has_k = true;
+        }
+      }
+
+      // Remember that we are just gathering the immediate inputs to the
+      // matmul, so there should be no prologue between a, b and the mul/sum.
+
+      // Check that the inputs have broadcasts that are not all in common, i.e.
+      // that there is at least one M and at least one N dimension.
+      TensorView* ltv = getTensorviewPriorToCast(bop->lhs()->as<TensorView>());
```

Comment:
Will do. BTW it's not just implicit upcasting from mul. If we have half precision inputs and a chain of operations in Float, then we perform this multiplication, if the segmenter segments just before the mul then it could insert casts at the segmentation edge (once we move past #1889).